Index deletes not applied when cluster UUID has changed

abeyad opened the pull request:
If a node was isolated from the cluster while a delete was happening, the node will ignore the deleted operation when rejoining as we couldn't detect whether the new master genuinely deleted the indices or it is a new fresh "reset" master that was started without the old data folder. We can now be smarter and detect these reset masters and actually delete the indices on the node if its not the case of a reset master.

Note that this new protection doesn't hold if the node was shut down. In that case it's indices will still be imported as dangling indices.

Closes #11665

abeyad commented:
@bleskes @dakrone @brwe feedback would be appreciated :)

dakrone commented on core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -152,46 +168,35 @@ public boolean indexMetaDataChanged(IndexMetaData current) {
         return true;
     }
 
-    public boolean blocksChanged() {
```
Please don't remove this method here, it would make this a breaking change in that it would break any plugins that relied on watching for ClusterChangeEvents where the blocks have changed

dakrone commented on core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -152,46 +168,35 @@ public boolean indexMetaDataChanged(IndexMetaData current) {
         return nodesDelta.removed();
     }
 
-    public boolean nodesAdded() {
```
same here, why did you remove `nodesAdded` and `nodesChanged`?

abeyad replied:
Good point, I only saw they weren't used in ES core so removed them to cleanup, but I'll put them back in (or verify plugins don't use them either, but it is probably not worth doing as part of this commit).

bleskes commented on core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -134,10 +142,18 @@ public boolean indexRoutingTableChanged(String index) {
     }
 
+    /**
+     * Returns <code>true</code> iff the metadata for the cluster has changed between
```
can we add a note saying this is an object level equality and not a true equal.

abeyad replied:
Done, I also added that note to every other reference equality check in the class

bleskes commented on core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java:
```diff
@@ -134,10 +142,18 @@ public boolean indexRoutingTableChanged(String index) {
     }
 
+    /**
+     * Returns <code>true</code> iff the {@link IndexMetaData} for a given index
```
same here re type of equality.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java:
```diff
@@ -0,0 +1,350 @@
+    }
+
+    // Tests that the indices change list is correct as well as metadata equality when the metadata has changed.
+    private static void metaDataChangesCheck(final boolean changeClusterUUID) {
```
I think we can fold this test into one test and rarely change the cluster uuid?

abeyad replied:
@bleskes I'm not clear on what is meant here?  The `testMetaDataChangesOnNoMasterChange` and `testMetaDataChangesOnNewClusterUUID` tests each use this helper method as the test is essentially the same except for whether we change the cluster UUID or not.  I figured we needed a test to check metadata changes when the cluster UUID is changed and one when it is not.

bleskes commented on core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java:
```diff
@@ -0,0 +1,350 @@
+    /**
+     * Test nodes removed check.
+     */
+    public void testNodesRemoved() {
```
can we test both node addition and removal (in one cluster state)?

abeyad replied:
Done, this method actually does also check node addition as well, I just didn't name it well because the `ClusterChangedEvent` only contains a `nodesRemoved` method, not a `nodesAdded` method.  So I changed the test method's name.

bleskes replied:
Does it checks the case where nodes are added and removed in one cluster state update? 

Also I'm confused - there is a `org.elasticsearch.cluster.ClusterChangedEvent#nodesAdded` ?

abeyad replied:
@bleskes Sorry you are correct, there is a `nodesAdded`.  What happened was, I originally took it out because I saw it was not used anywhere in core, until @dakrone pointed out that they could be used by plugins and shouldn't be removed.  So I will add a test in here for `nodesAdded` as well (all part of the same test) and include a scenario where nodes are both added and removed in the same cluster state update.

abeyad replied:
Also, I'll include a test for the `nodesChanged` method (as part of the same test).

bleskes commented:
I left some very minor comments. Looks good. I thought we had a test somewhere about index deletion with master change (which we should be able to enable now), but maybe I'm wrong. @brwe do you remember?

bleskes commented:
also re the pr description:

> This commit fixes the issue by only reimporting indices from data nodes if and only
if the cluster UUID on the master node is different from the cluster
UUID of the previous cluster state on the data node

This inaccurate - on disk dangling indices will still be imported - for example if a node was off while the index was deleted. This change only deals with the case where a node was isolated from the cluster while a delete happened (but was up!) and where a master failure happened during a delete operation (but after the delete was already  committed).   I think we should something like:  

If a node was isloated from the cluster while a delete was happening, the node will ignore the deleted operation when rejoining as we couldn't detect whether the new master genuinely deleted the indices or is a new fresh "resetted" master that was started without the old data folder. We can now be smarter and detect these resetted masters and actually delete the indices if this is not the case. 

Note that this new protection doesn't hold if the node was shut down. In that case it's indices will still be imported as dangling indices.

brwe commented:
> @brwe do you remember?

I think that is the one that @abeyad enabled here: https://github.com/elastic/elasticsearch/pull/16825/files#diff-c454159f60a0127854028ccb5502500eL1076 I know of no other one.

bleskes commented:
@brwe I missed that one line in the white space fixing. Jet lag... :( thanks!

abeyad commented:
I fixed the issues raised by @bleskes and @dakrone and all tests pass.  I also updated the PR description and will change the commit message as well once the PR is approved.

bleskes commented:
LGTM. Thanks @abeyad
